trace-client/src/main/java/org/hypertrace/trace/client/AvroBackedValueSource.java

```diff
@@ -0,0 +1,25 @@
+package org.hypertrace.trace.client;
+
+import java.util.Optional;
+import javax.annotation.Nullable;
+import org.hypertrace.core.datamodel.AttributeValue;
+import org.hypertrace.core.datamodel.Attributes;
+import org.hypertrace.core.datamodel.MetricValue;
+import org.hypertrace.core.datamodel.Metrics;
+
+abstract class AvroBackedValueSource implements ValueSource {
+
+  Optional<String> getAttributeString(@Nullable Attributes attributes, String key) {
+    return Optional.ofNullable(attributes)
+        .map(Attributes::getAttributeMap)
```

Comment:
what if map is empty?

Reply:
If either the map or the result of `.get` return null, optional handles that and converts to an empty optional: 
```java
    public <U> Optional<U> map(Function<? super T, ? extends U> mapper) {
        Objects.requireNonNull(mapper);
        if (!isPresent()) {
            return empty();
        } else {
            return Optional.ofNullable(mapper.apply(value));
        }
    }
```